File: stacks/python/base/code/queue.tpl.py
"""
File: queue.tpl.py
Purpose: Background job processing with Celery
Generated for: {{PROJECT_NAME}}
"""

from celery import Celery
from celery.schedules import crontab
from typing import Any, Callable, Optional
from functools import wraps
import logging

logger = logging.getLogger(__name__)


def create_celery_app(
    name: str,
    broker_url: str = "redis://localhost:6379/0",
    result_backend: str = "redis://localhost:6379/1",
) -> Celery:
    """Create and configure Celery application"""
    
    app = Celery(
        name,
        broker=broker_url,
        backend=result_backend,
    )

    app.conf.update(
        task_serializer="json",
        accept_content=["json"],
        result_serializer="json",
        timezone="UTC",
        enable_utc=True,
        task_track_started=True,
        task_time_limit=30 * 60,  # 30 minutes
        task_soft_time_limit=25 * 60,  # 25 minutes
        worker_prefetch_multiplier=1,
        task_acks_late=True,
        task_reject_on_worker_lost=True,
    )

    return app


# Default app instance
celery_app = create_celery_app("{{PROJECT_NAME}}")


# Task decorators
def task(
    name: Optional[str] = None,
    max_retries: int = 3,
    retry_backoff: bool = True,
    queue: str = "default",
):
    """Custom task decorator with retry logic"""
    def decorator(func: Callable) -> Callable:
        @celery_app.task(
            name=name or func.__name__,
            bind=True,
            max_retries=max_retries,
            default_retry_delay=60,
            queue=queue,
        )
        @wraps(func)
        def wrapper(self, *args, **kwargs) -> Any:
            try:
                return func(*args, **kwargs)
            except Exception as exc:
                logger.error(f"Task {self.name} failed: {exc}")
                if retry_backoff:
                    raise self.retry(exc=exc, countdown=2 ** self.request.retries * 60)
                raise self.retry(exc=exc)

        return wrapper
    return decorator


# Example tasks
@task(name="send_email", queue="notifications")
def send_email(to: str, subject: str, body: str) -> dict:
    """Send an email"""
    logger.info(f"Sending email to {to}: {subject}")
    # Implement email sending logic
    return {"status": "sent", "to": to}


@task(name="process_file", queue="processing")
def process_file(file_path: str) -> dict:
    """Process a file in the background"""
    logger.info(f"Processing file: {file_path}")
    # Implement file processing logic
    return {"status": "processed", "file": file_path}


# Scheduled tasks (beat schedule)
celery_app.conf.beat_schedule = {
    "cleanup-expired-tokens": {
        "task": "cleanup_tokens",
        "schedule": crontab(hour=0, minute=0),  # Daily at midnight
    },
    "send-daily-digest": {
        "task": "send_digest",
        "schedule": crontab(hour=8, minute=0),  # Daily at 8 AM
    },
}


# Usage:
# Start worker: celery -A queue worker -l INFO -Q default,notifications,processing
# Start beat: celery -A queue beat -l INFO
#
# Send task:
# from queue import send_email
# send_email.delay("user@example.com", "Hello", "World")

File: stacks/node/base/code/queue.tpl.js
/*
File: queue.tpl.js
Purpose: Background job processing with BullMQ
Generated for: {{PROJECT_NAME}}
*/

const { Queue, Worker, QueueScheduler } = require('bullmq');
const Redis = require('ioredis');

// Redis connection for all queues
const connection = new Redis({
    host: process.env.REDIS_HOST || 'localhost',
    port: parseInt(process.env.REDIS_PORT || '6379', 10),
    maxRetriesPerRequest: null,
});

// Queue registry
const queues = {};
const workers = {};

function createQueue(name, options = {}) {
    if (queues[name]) {
        return queues[name];
    }

    const queue = new Queue(name, {
        connection,
        defaultJobOptions: {
            attempts: 3,
            backoff: {
                type: 'exponential',
                delay: 1000,
            },
            removeOnComplete: {
                age: 24 * 60 * 60, // 24 hours
                count: 1000,
            },
            removeOnFail: {
                age: 7 * 24 * 60 * 60, // 7 days
            },
            ...options.defaultJobOptions,
        },
        ...options,
    });

    queues[name] = queue;
    return queue;
}

function createWorker(queueName, processor, options = {}) {
    if (workers[queueName]) {
        return workers[queueName];
    }

    const worker = new Worker(
        queueName,
        async (job) => {
            console.log(`Processing job ${job.id} in queue ${queueName}`);
            try {
                const result = await processor(job.data, job);
                console.log(`Job ${job.id} completed`);
                return result;
            } catch (error) {
                console.error(`Job ${job.id} failed:`, error);
                throw error;
            }
        },
        {
            connection,
            concurrency: options.concurrency || 5,
            ...options,
        }
    );

    worker.on('completed', (job) => {
        console.log(`Job ${job.id} completed successfully`);
    });

    worker.on('failed', (job, error) => {
        console.error(`Job ${job?.id} failed:`, error.message);
    });

    workers[queueName] = worker;
    return worker;
}

// Pre-defined queues
const emailQueue = createQueue('email');
const processingQueue = createQueue('processing');

// Job handlers
const jobHandlers = {
    sendEmail: async (data) => {
        const { to, subject, body } = data;
        console.log(`Sending email to ${to}: ${subject}`);
        // Implement email sending
        return { status: 'sent', to };
    },

    processFile: async (data) => {
        const { filePath } = data;
        console.log(`Processing file: ${filePath}`);
        // Implement file processing
        return { status: 'processed', file: filePath };
    },
};

// Helper to add jobs
async function addJob(queueName, jobName, data, options = {}) {
    const queue = queues[queueName] || createQueue(queueName);
    return queue.add(jobName, data, options);
}

// Scheduled/recurring jobs
async function scheduleRecurringJobs() {
    // Daily cleanup at midnight
    await processingQueue.add(
        'cleanupExpiredTokens',
        {},
        {
            repeat: {
                pattern: '0 0 * * *', // Daily at midnight
            },
        }
    );

    // Daily digest at 8 AM
    await emailQueue.add(
        'sendDailyDigest',
        {},
        {
            repeat: {
                pattern: '0 8 * * *', // Daily at 8 AM
            },
        }
    );
}

// Initialize workers
function startWorkers() {
    createWorker('email', jobHandlers.sendEmail);
    createWorker('processing', jobHandlers.processFile);
    console.log('Workers started');
}

// Graceful shutdown
async function shutdown() {
    console.log('Shutting down workers...');
    await Promise.all(Object.values(workers).map((w) => w.close()));
    await Promise.all(Object.values(queues).map((q) => q.close()));
    await connection.quit();
}

process.on('SIGTERM', shutdown);
process.on('SIGINT', shutdown);

module.exports = {
    createQueue,
    createWorker,
    addJob,
    emailQueue,
    processingQueue,
    startWorkers,
    scheduleRecurringJobs,
};

// Usage:
// const { addJob, startWorkers } = require('./queue');
// startWorkers();
// await addJob('email', 'sendEmail', { to: 'user@example.com', subject: 'Hello', body: 'World' });
